### PR TITLE
Revise the use of nonces in requests to store a URL Metric and block cross-origin requests

### PR DESCRIPTION
**Pull request by westonruter**
This is the first PR to address #1496.

1. Eliminate the use of the `wp_rest` nonce when making POST requests to store a URL Metric. Since the endpoint does not require authentication, the nonce adds no value.
2. Add enforcement of the `Origin` request header so that it must match `home_url()` in order for the URL Metric storage POST request to be accepted. The WordPress REST API by default allows for cross-origin requests, but this makes no sense specifically for storing a URL Metric since the client must by definition be on the frontend of the site.
3. Eliminate the use of `wp_create_nonce()` to compute the HMAC to validate that the provided query-vars `slug` actually corresponds to the provided `url` in the request (that is, the URL of the current page). There is no need to use `wp_create_nonce()` here, since there does not need to be an expiration and it does not need to be tied to the current logged-in user.

By eliminating the use of nonces, clients will not have issues when submitting a URL Metric for storage when a page cache is holding onto the rendered page for longer than a day.

### CODE REVIEW

File: plugins/optimization-detective/storage/data.php [hunk 1]
```diff
@@ -141,42 +141,45 @@ function od_get_url_metrics_slug( array $query_vars ): string {
 }
 
 /**
- * Computes nonce for storing URL Metrics for a specific slug.
+ * Computes HMAC for storing URL Metrics for a specific slug.
  *
  * This is used in the REST API to authenticate the storage of new URL Metrics from a given URL.
  *
  * @since 0.1.0
+ * @since n.e.x.t Renamed from od_get_url_metrics_storage_nonce().
```
Comment (felixarntz) [inline plugins/optimization-detective/storage/data.php]: Better to consider this an entirely new function. I would argue it wasn't "just" renamed, the purpose was also changed as it now returns a different kind of value.

File: plugins/optimization-detective/storage/data.php [hunk 2]
```diff
@@ -141,42 +141,45 @@ function od_get_url_metrics_slug( array $query_vars ): string {
...
+	$action = "store_url_metric:$slug:$url";
+	return wp_hash( $action, 'nonce' );
 }
 
 /**
- * Verifies nonce for storing URL Metrics for a specific slug.
+ * Verifies HMAC for storing URL Metrics for a specific slug.
  *
  * @since 0.1.0
+ * @since n.e.x.t Renamed from od_verify_url_metrics_storage_nonce().
```
Comment (felixarntz) [inline plugins/optimization-detective/storage/data.php]: See above, I'd prefer we consider this a new function.

File: plugins/optimization-detective/storage/rest-api.php [hunk 3]
```diff
@@ -88,6 +88,42 @@ function od_register_endpoint(): void {
...
+ * @since n.e.x.t
+ * @access private
+ *
+ * @see get_allowed_http_origins()
+ * @see is_allowed_http_origin()
+ *
+ * @param string $origin Origin to check.
+ * @return bool Whether the origin is allowed.
+ */
+function od_is_allowed_http_origin( string $origin ): bool {
```
Comment (felixarntz) [inline plugins/optimization-detective/storage/rest-api.php:106]: Why can't we use `is_allowed_http_origin()` directly instead of this function?

If it's just about the port support, I'd say that's an edge case so then I think we should at least use the original function anytime there's no port, and the custom logic only if there is one.

I'd also say we could simplify it: If there is a port in the passed `$origin`, simply remove it for the comparison and call `is_allowed_http_origin()` with that. It's not really that important to validate that the port also matches IMO, especially because of how much of an edge case that is.

**Review by felixarntz — APPROVED** [review]
@westonruter This looks great! Just some minor feedback, which would be good to address, but no blockers.

File: plugins/optimization-detective/storage/data.php (hunk 1 above)
Comment (westonruter) [reply plugins/optimization-detective/storage/data.php]: Sure. Done in 9d6e80784

File: plugins/optimization-detective/storage/data.php (hunk 2 above)
Comment (westonruter) [reply plugins/optimization-detective/storage/data.php]: Sure. Done in 9d6e80784

File: plugins/optimization-detective/storage/rest-api.php (hunk 3 above)
Comment (westonruter) [reply plugins/optimization-detective/storage/rest-api.php:106]: Well, unit tests run the home being `localhost:8889` so it's not an edge case from that perspective. 

I've gone ahead and just stripped out the port number in 545f0c3e6.

Comment (felixarntz) [reply plugins/optimization-detective/storage/rest-api.php:106]: Great, I think that simplifies the implementation, and validating the same port number is probably not super important (for `localhost` sites there's probably not too much of a security implication anyway).
